### DISCUSSION TITLE
Parallel State Transitions

### DISCUSSION THREAD

**Discussion by mashtonian** [nsk90/kstatemachine#42] [Q&A] [answered]
I can't tell from the wiki how transitions are fired for parallel states.

If the machine is "in" multiple parallel states and they all have transitions that match an event, are they all fired? If so, in what order? Sequentially or multithreaded?

I think my project has a use case for parallel states but it depends on the above.

**Comment by nsk90** [comment, marked as answer]
As I see from test case "process event by parallel states negative", multiple transitions for one event is currently not possible. Exception will be thrown in this case.

The rule of firing neighbour notifications in all the library: such notifications are fired sequentially, without multithreading, execution order is not specified, you should not rely on it.

**Reply by nsk90** [reply]
How do you want to use multiple transitions in parallel states? 
Current limitation takes place, as those transitions may affect each other by changing machine in conflicting way.
Looks that the only case when it may work when all such transitions are targeted to states that are deeper children of "parallel" layer state.

**Comment by mashtonian** [comment]
Fair enough.

I'll have a play around and see what is possible.

**Comment by mashtonian** [comment]
I am writing a snooker scoreboard app.

In snooker, if the game is in the state "on a red" and also the state "free
ball" then the event "pot red" adds 1 to the score but doesn't decrease the
remaining reds count. If it is not a free ball then the remaining red count
is decremented.

You can also be in a "free ball" state when in the "on a colour" state.
Hence why I am considering parallel states.

In essence, there are small semantic side effect differences for some
transitions when in a free ball situation. All other transitions remain
unaffected.

I could, of course, simply keep "free ballness" as a class variable but
prefer the cleanliness of using StateMachine.

I prefer to use parallel states, keeping the transition side effects in the
master "on a red" state, but with some logic that tests whether the free
ball state is active or not, to adjust the side effects.

Now I think about it, it's also unclear to me how the machine would exit
the free ball state. Are all parallel states exited when any of them exit?

**Reply by nsk90** [reply]
Parallel states are exit when there parent state exit, so they all exit at the same time.

**Reply by nsk90** [reply]
I think parallel states fits your case good. 

```
I prefer to use parallel states, keeping the transition side effects in the
master "on a red" state, but with some logic that tests whether the free
ball state is active or not, to adjust the side effects.
```
Maybe you need to run in parallel couple of states like: freeball true/ freeball false.  And switch between them. Other way I dont see how you are going to exit single state.

See this sample, maybe will help
https://github.com/nsk90/android-kstatemachine-sample/blob/main/app/src/main/java/ru/nsk/kstatemachinesample/ui/main/MainViewModel.kt
